### ISSUE TITLE
Regression in CSP handling causes blank screen

### ISSUE THREAD

**Issue by christianlupus** [nextcloud/richdocuments#2065]
## Bug description
**Describe the bug**
Recently I updated the app to v5.0.2 and Collabora did no longer work. I have no clear idea when the issue arose as I am not using the app on a daily basis. 

This might be related or even a duplicate of #2052, #2058, and #2060.

**To Reproduce**
1. Navigate in the NC web interface to a location with e.g. an ODT file.
2. Open the file by clicking it

For more details see also below.

**Expected behavior**
The Collabora editor is loading and allows me to edit the file.

**Screenshots**
The screen keeps mostly blank as depicted here:
![grafik](https://user-images.githubusercontent.com/8202922/154795863-eb095686-8a59-4164-a5e2-7daf526649a1.png)

## Additional information and research

I looked into the issue a bit more. The problem is that the CSP prevents the browser from opening the `iframe` with the Collabora editor.

I made a `git bisect` to get the failing commit a5b345f.

To understand my setup, I have one machine running an Apache reverse proxy and a docker-compose that contains all parts of the installation (DB, Redis, cron, NC server, and Collabora containers). The NC server and the Collabora are separate containers and exported on individual ports to be accessed by the Apache reverse proxy. The containers are published on two virtual hosts `nextcloud.example.com` and `collabora.example.com`.
There is another NC installation with a similar setup that uses this Collabora installation as well. This is the reason for splitting out the subdomain of Collabora.

I will keep the git repo in place for the primary instance to be able to do a quick checkout/debug if was required. Feel free to contact me here.

## Client details:
 - OS: Archlinux
 - Browser: Firefox
 - Version 96.0.2
 - Device: desktop

## Server details

**Operating system**: Linux/Docker images

**Web server:** Docker image with Apache reverse proxy

**Database:** MariaDB

**PHP version:** 8.0.12

**Nextcloud version:** 22.2.3

**Version of the richdocuments app** 4.2.4

**Version of Collabora Online** Current docker image from 2022-02-18

<details>
<summary>Logs</summary>

#### Nextcloud log (data/nextcloud.log)
I do not see anything obvious in the backend log.

#### Browser log
In the console a message like the following is printed when using a bad commit:
```
Content Security Policy: Die Einstellungen der Seite haben das Laden einer Ressource auf https://collabora.example.com/browser/e4d06c2/cool.html?WOPISrc=https%3A%2F%2Fnextcloud.example.com%2Fnextcloud%2Findex.php%2Fapps%2Frichdocuments%2Fwopi%2Ffiles%2FABCD&title=Kooperation.odt&lang=de&closebutton=1&revisionhistory=1 blockiert ("form-action").
```
I replaced the real domain and removed some alphanumeric values by `ABCD` as I expect these are somehow a token.

I get the following CSR headers from the NC server in the browser when opening URLs of the form `https://nextcloud.example.com/nextcloud/apps/richdocuments/index?fileId=12345` for the good commits:
```
Content-Security-Policy: default-src 'none';base-uri 'none';manifest-src 'self';script-src 'nonce-BASE64_STRING' blob: 'unsafe-inline';style-src 'self' * 'unsafe-inline';img-src 'self' data: blob: * cid: https://*.tile.openstreetmap.org;font-src 'self' data: *;connect-src 'self' blob: stun.nextcloud.com:443 stun.example.com:3478;media-src 'self' blob:;frame-src 'self' data: https://collabora.example.com prezi.com player.vimeo.com vine.co www.youtube.com;child-src 'self' blob:;frame-ancestors 'self';worker-src 'self' blob:;form-action 'self' https://collabora.example.com
```

The bad headers are
```
Content-Security-Policy: default-src 'none';base-uri 'none';manifest-src 'self';script-src 'nonce-BASE64_STRING' blob: 'unsafe-inline';style-src 'self' * 'unsafe-inline';img-src 'self' data: blob: * cid: https://*.tile.openstreetmap.org;font-src 'self' data: *;connect-src 'self' blob: stun.nextcloud.com:443 stun.example.com:3478;media-src 'self' blob:;frame-src 'self' data: prezi.com player.vimeo.com vine.co www.youtube.com https://collabora.example.com;child-src 'self' blob:;frame-ancestors 'self';worker-src 'self' blob:;form-action 'self'
```

I see (apart from the ordering) the difference in the `form-action` part. 
I am not sure if the syntax is correct (is a space before `;` required?). But I suspect this to be no issue as it might have caused other issues in the backend quite earlier.
</details>

**Comment by christianlupus** [comment]
I investigated a bit more and found a workaround at least. I had to tweak the CSP headers as follows:
```
Header always edit Content-Security-Policy ";\s*form-action\s+([^;]*)" ";form-action $1 https://collabora.example.com"
```

This will append the Collabora domain to the `form-action` directive. It might help to check if this is done by the code of the app.

**Comment by juliushaertl** [comment]
Thanks a lot for the in-depth investigation. Interestingly I cannot reproduce this issue on any of the setups I have access to. 

Could you maybe check if the following patch also addresses the issue for you? https://github.com/nextcloud/richdocuments/pull/2051/files

In addition I'd be quite curious about why the initial CSP rule does not apply here, since you mention that you have left one node for debugging, could you maybe also apply the following patch there and try to get the log output for opening a document with that?

```
diff --git a/lib/AppInfo/Application.php b/lib/AppInfo/Application.php
index 3d640340..4bd16736 100644
--- a/lib/AppInfo/Application.php
+++ b/lib/AppInfo/Application.php
@@ -201,6 +201,7 @@ class Application extends App implements IBootstrap {
                // Do not apply CSP rules on WebDAV/OCS
                // Ideally this could be a middleware running after the controller execution before rendering the result to only do it on page response
                if ($container->getServer()->getRequest()->getScriptName() !== '/index.php') {
+                       \OC::$server->getLogger()->error('Not applying CSP for ' . $container->getServer()->getRequest()->getScriptName());
                        return;
                }
 ```

**Comment by christianlupus** [comment]
I applied the patch as provided and looked at `data/nextcloud.log` while clicking on an ODT file to open. I kept the workaround in my Apache reverse proxy to update the CSP header (postprocessing). I got a few lines logged that did contain the string `Not applying CSP for`:
```
{"reqId":"hEovRMOHNnpt1cyb8E6D","level":3,"time":"2022-02-21T14:34:10+00:00","remoteAddr":"134.96.204.3","user":"christian-nextcloud","app":"no app in context","method":"GET","url":"/nextcloud/ocs/v2.php/apps/notifications/api/v2/notifications","message":"Not applying CSP for /nextcloud/ocs/v2.php","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0","version":"22.2.3.0"}
{"reqId":"X2Z4wjbCxhfizy9bsaiz","level":3,"time":"2022-02-21T14:34:14+00:00","remoteAddr":"134.96.204.3","user":"christian-nextcloud","app":"no app in context","method":"PUT","url":"/nextcloud/apps/user_status/heartbeat","message":"Not applying CSP for /nextcloud/index.php","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0","version":"22.2.3.0"}
{"reqId":"vOPkZuAtiQCcsrSS1SDs","level":3,"time":"2022-02-21T14:34:22+00:00","remoteAddr":"134.96.204.3","user":"christian-nextcloud","app":"no app in context","method":"PUT","url":"/nextcloud/apps/user_status/heartbeat","message":"Not applying CSP for /nextcloud/index.php","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0","version":"22.2.3.0"}
{"reqId":"4VArYttFHJrUEHXxkXzT","level":3,"time":"2022-02-21T14:34:24+00:00","remoteAddr":"134.96.204.3","user":"christian-nextcloud","app":"no app in context","method":"GET","url":"/nextcloud/apps/richdocuments/index?fileId=3917053&requesttoken=baz","message":"Not applying CSP for /nextcloud/index.php","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0","version":"22.2.3.0"}
{"reqId":"khiE9fCgS3x6dDpbQPBN","level":3,"time":"2022-02-21T14:34:25+00:00","remoteAddr":"172.21.0.4","user":"--","app":"no app in context","method":"GET","url":"/nextcloud/index.php/apps/richdocuments/wopi/files/3917053_bar?access_token=foo&access_token_ttl=0","message":"Not applying CSP for /nextcloud/index.php","userAgent":"COOLWSD HTTP Agent 21.11.2.2","version":"22.2.3.0"}
{"reqId":"wAnfo4G5BllFAUiEMoRY","level":3,"time":"2022-02-21T14:34:25+00:00","remoteAddr":"172.21.0.4","user":"--","app":"no app in context","method":"GET","url":"/nextcloud/index.php/apps/richdocuments/wopi/files/3917053_bar/contents?access_token=foo&access_token_ttl=0","message":"Not applying CSP for /nextcloud/index.php","userAgent":"COOLWSD HTTP Agent 21.11.2.2","version":"22.2.3.0"}
{"reqId":"9gSeqDGTJaiP0kV0dVUT","level":3,"time":"2022-02-21T14:34:27+00:00","remoteAddr":"134.96.204.3","user":"christian-nextcloud","app":"no app in context","method":"GET","url":"/nextcloud/apps/notes/notes/2698089","message":"Not applying CSP for /nextcloud/index.php","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Biscuit/1.2.22 Chrome/91.0.4472.164 Electron/13.1.9 Safari/537.36","version":"22.2.3.0"}
{"reqId":"p1GZi45u0QkcenzkqQEg","level":3,"time":"2022-02-21T14:34:27+00:00","remoteAddr":"134.96.204.3","user":"christian-nextcloud","app":"no app in context","method":"GET","url":"/nextcloud/avatar/christian-nextcloud/32","message":"Not applying CSP for /nextcloud/index.php","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0","version":"22.2.3.0"}
```

This was tested with 4.2.4 and no other patched/changes were applied. If you need the complete log (not just `tail -f nextcloud.log | grep 'Not applying CSP for'`), feel free to ask.

Regarding the linked PR/patch #2051, I have to check if I can backport to 4.2.4 as the 5.x.x is not compatible with server version 22.

**Comment by christianlupus** [comment]
I was able to cherry-pick the commit 7e20600 in 4.2.4 and test with that in place. That will add the required field to the CSP header. So, for 5.0.3 it will be solved. For server version 22 you would need a backport.

Please tell me, if there is any more need fo the testing instance. Otherwise, I could remove the repo from the production system and switch back to the releases installed by the server. Also, I would update to server version 23 if the test instance on v22 is not needed any longer. Please give me acknowledgment, @juliushaertl. Thanks!

**Comment by maxxer** [comment]
I'm still getting empty screen even after applying 7e20600. I get a warning in the JS console for malformed CSP rule.

The `form-action` rule contains the dot-escaped url of my Collabora instance (`https://abc\.domain\.com`)

**Comment by christianlupus** [comment]
To make sure, it is syntactically correct, I commented out my workaround in the reverse proxy configuration. It works with the patch applied. For me, there is no dot-escaping in place.

**Comment by mario-spitze** [comment]
Same problem at my environment. The setup is Nextcloud 22.2.5 and 22.2.3 (test both), Richdocuments  app 4.2.4, Collabora Code 21.11.1.4.1 on Docker with different domain and everything behind Apache reverse proxy.

Use the patch to change the two lines solves the problem.

**Comment by ironicx** [comment]
I had the same CSP "form-action"error with richdocuments 5.0.2, nextcloud 23.0.2 and Collabora on a different subdomain. Applying the patch https://github.com/nextcloud/richdocuments/commit/7e20600755078d8c44dbde1daf9fa2f9f956f04f fixed it for me. Thanks :)

**Comment by juliushaertl** [comment]
Closing as fixed with #2051 and released as 5.0.3 as mentioned above.

**Comment by maxxer** [comment]
I've updated to 5.0.3, fixed the malformed CSP line by deleting a stale row in config, but I still have 

```
Refused to send form data to 'https://collabora.domain.com/browser/e4d06c2/cool.html?WOPISrc=https%3A%2F%2Fcloud.domain.com%2Fnc%2Findex.php%2Fapps%2Frichdocuments%2Fwopi%2Ffiles%2F1483828_5236c4bbc9e9d&title=autosave.ods&lang=en&closebutton=1&revisionhistory=1' because it violates the following Content Security Policy directive: "form-action 'self'".
```

Our NC is at `cloud.domain.com/nc`
